### ISSUE TITLE
Multiple `require`, `require.ensure` and `require.resolve` failures when dynamically requiring chunks in version 1 and 2

### ISSUE THREAD

**Issue by wyuenho** [webpack/webpack#2597]
**Update**: https://github.com/webpack/webpack/issues/2597#issuecomment-223771395

Test cases: https://github.com/wyuenho/webpack-require-chunk-as-module-fail

The way Webpack works is there's usually one entry chunk with the Webpack runtime included. The runtime has a JSONP callback called `webpackJsonp`. Each chunk, when loading has completed, will call the Jsonp function, which will install the chunk and the extra modules. The Jsonp function will execute the chunk as if it were a module if it is also an entry chunk (distinguished by a module with an ID 0). 
This implies that all the chunks will have to be entry chunks and we can't use the CommonChunkPlugin. 

This limitation totally defeats the purpose of code-splitting.

To get around this, we'll need to have the ability to resolve to Webpack's internal ID in order to use `__webpack_require__` to require it at runtime (or better yet, just make `require(lateBoundModuleName)` work). Unfortunately, Webpack doesn't treat the chunks as requirable modules (unless it is the one and only entry chunk, in which Webpack will require it internally automatically), and it doesn't inject a map of normal node module paths to Webpack's internal module IDs into the runtime.

In short, Webpack's implementation of node style require is incomplete and broken. The missing stuff is what we need.

We cannot use Webpack at this point without some significant effort to alter it, the documentation is too disparate, and the code it too difficult to follow. If you can keep us some pointers, we might be able to make a PR.

**Comment by taurose** [comment]
I'm confused about what you're trying to achieve. What's the `fetch` for? Why do you need `require.resolve` and multiple entry chunks?

Does this help? https://github.com/webpack/webpack/issues/300#issuecomment-45313650

**Comment by wyuenho** [comment]
@taurose I've replace `fetch` with `promisescript`. I forgot fetch doesn't exec the loaded script, I just used it to make the bundle smaller.

Again, ideally we'd like to make chunks act like normal requirable modules. The way Webpack works is, in order to exec the chunks, the chunks must be entry chunks, which essentially just moved the entry point from one place to another, the dep graph is essentially the same.

We'd like the ability to have Webpack generate the chunks, use the `CommonChunkPlugin` to move the runtime and common deps to the sole initializing entry chunk, and inside that initializing chunk, we'll load the rest of the chunks on demand, and require them as soon as they are loaded.

**Comment by sokra** [comment]
I'm also pretty confused by your question, but it looks like you just want to load some code on demand. If this is what you want to do:

``` js
// no fetch, no promisescript, only index.js as entry point
// in index.js
require(["./a"], function(a) {
  // do something with a
});
```

**Comment by wyuenho** [comment]
I've completely updated the test cases and the comments in the code again, incorporating all of your comments.

The reason I can't use `require.ensure` is it's broken for different reasons in both Webpack 1 and 2.
For my purpose, the module name is not known in advance, it must be resolved at runtime, so I cannot require statically with `require.ensure`, and both versions of Webpacks fails at that.

I just tested https://github.com/webpack/webpack/issues/300#issuecomment-45313650, it appears to work for Webpack 1 using `promisescript`, this might fit my purpose. However, there are still at least 3-4 different issues related to this scenario reported. Please check it out and experiment with them.

Lastly, I'd really wish the error messages and documentation were clearer on the `[]` bug in Webpack 1.

**Comment by taurose** [comment]
Are all the dynamic dependencies bundled together with the main bundle like in your test cases, or do you have to bundle them separately?

**Comment by wyuenho** [comment]
I just committed the erroneously ignored `a` bundle for the webpack2 case. For the webpack1/good webpack2/bad1 cases, there's only 1 bundle created. The rest create multiple bundles. My real app creates separate bundles for the dynamic deps.

**Comment by taurose** [comment]
If all the deps are available during compile-time you should be able to use `require.ensure`:

```js
var a = 'a';
require.ensure([], function (require) {
   // all dynamic entry modules inside /dynamic/ folder
   console.log(require('./dynamic/' + a));
});
```

This will only create one chunk for all deps though. For one chunk per module, you can use `require.context` in combination with bundle-loader or promise-loader (suggested [here](https://github.com/webpack/webpack/issues/2029#issuecomment-182818697)).

````js
var a = './a';
require.context('bundle!./dynamic')(a)(function(moduleA){
	console.log(moduleA);
});
```

If these dependencies need to be bundled independently though, I don't think you can use webpack chunks for that. Instead, you could build the dependencies as stand-alone libraries that export to `window` and then use `promisescript` to load them.

**Comment by wyuenho** [comment]
I checked out the suggestions you've linked to and updated the code with more test cases again.

The one thing you have to beware of is all the chunks are built into the the same output directory. Webpack is not make, it doesn't support recursive build. In Webpack 1, in order to use `require.context`, I have to run the build **twice** - the first one to build the chunks that will be dynamically required, and the second one to have Webpack resolve the dependencies that's in the **output** directory.

`System.import` in Webpack 2 also fails miserably with a parsing bug and 2 resolve bugs.

Let me make this clear, when I have a dep graph like this:
```
- src/
  - index.js
  - a.js
```

```javascript
// index.js
var a = './a';
require.resolve(a);
```
Even if I configured `a` to live in a different chunk in the config, it should just **work**. `require.resolve()` should be able to fetch, exec, install and resolve the chunk `a` is in automatically. All of these weird contortions like `require.context`, `require.ensure`, `__webpack__*` are there because Webpack tries too hard to resolve all dependencies statically. Webpack just refuses to inject a dependency graph into the entry chunk under most cases, thereby preventing a large class of programming idioms from being use, because *cannot* resolve dependencies on demand reliably.

Webpack is a compiler, not an interpreter, in order to behave like one, as least for import/exports, you'll need to support late binding of modules in the runtime.

**Comment by sokra** [comment]
webpack is a static module bundler. All modules **MUST** be resolved statically. It doesn't support resolving on demand. It can resolve some dynamic expression statically (`require("./folder/"+ a)`), but they are still resolved statically. You could also use `require(a)`, but you must provide the required information for statical resolving.

You may are looking for a module loader (and not a module bundler). A module loader (i. e. require.js) resolves modules on runtime.

> It's preventing a large class of programming idioms from being used, because cannot resolve dependencies on demand reliably.

I don't think this class is soo large...

btw. `require(a)` in a browser environment is just not possible, because to resolve and load the request the browser need to do async network requests, which are not possible in a sync function call. (sync xhr is bad)

**Comment by wyuenho** [comment]
I'm actually looking for a static module bundler that allows me to split the bundles and be able to use them effectively. What I'd like to see is when I've loaded the chunks down, using whatever mechanism I choose (ideally System.import), the modules in the chunks should be made available and requirable from the point of view of the piece of code that initiated the load.

Since Node style `require`s may contain lots of relative paths, to enable true dynamic requires, you need something like Git's tree object, which is just a snapshot of the directory tree. Without it, require doesn't know the calling module is sitting in the directory tree, and where to find the require request. If Webpack had this in the entry chunk, it wouldn't need most of these weird context and resolve configs and module APIs.

The programmer should be responsible to make sure the chunks loaded will have the module that's needed, which means he has to configure the split points correctly and structure the code to support this. Webpack should be responsible for transparently resolving the dependencies as expected.

I think it's possible emulate a synchronous function call in JS using generators, and it's possible to transpile generators to ES5 code using Facebook's regenerator. Inside the `require`, and `require.resolve` implementations, you can just call the inner generator's `send` method to suspend until XHR has returned with some data. This should give you good synchronous looking code without actually busy waiting.

**Comment by dpundir** [comment]
Have we got solution for the dynamic require through webpack problem?
I too have similar requirement:
require([varname], function() {}) OR
require(["myfolder/" + locale + "/folder"], function() {})

**Comment by maximoleinyk** [comment]
Hi guys,

Could someone please help me to rewrite this function https://github.com/maximoleinyk/reactjs/blob/master/src/js/app/app.jsx#L42 so I could see separate bundles for each module using dynamic require.

I still don't understand how to use `require.context` function in my case. I'm happy to specify static information in webpack.config.js file just don't know how to do that.

Any help would be appreciated!

**Comment by maximoleinyk** [comment]
I found a solution 

```

	getChildRoutes(state, callback) {
		this.config.modules.forEach((name) => {
			if (!location.pathname.startsWith('/page/' + name)) {
				return;
			}

			require(`bundle!app/${name}/routes`)((module) => {
				callback(null, module.default);
			});
		});
	}
```

Cheers!

**Comment by TheLarkInn** [comment]
This is resolved.
